apps/design/shared/src/layout.ts

```diff
@@ -787,32 +1146,105 @@ function layOutBallotHelper(
+    while (
+      heightUsed < gridHeight(contestsRowHeight) &&
+      contestSectionsLeftToLayOut.length > 0
+    ) {
+      const contestSection = assertDefined(contestSectionsLeftToLayOut.shift());
+      // Lay out contests just to get their heights
+      const contestsWithHeights = contestSection.map((contest) => {
+        const ContestComponent =
+          contest.type === 'candidate' ? CandidateContest : BallotMeasure;
+        const [{ height }] = ContestComponent({
+          election,
+          contest,
+          row: 0,
+          gridRow: 0,
+          gridColumn: 0,
+          pageNumber: 0,
+        });
+        if (height > gridHeight(MAX_CONTEST_ROW_HEIGHT)) {
+          throw new Error(`Contest ${contest.id} is too tall to fit on a page`);
+        }
+        return {
+          contest,
+          height: height + gridHeight(CONTEST_ROW_MARGIN),
+        };
+      });
+
+      const { columns, height, leftoverElements } = layOutInColumns({
+        elements: contestsWithHeights,
+        numColumns: contestSection[0].type === 'candidate' ? 3 : 1,
+        maxColumnHeight: gridHeight(contestsRowHeight) - heightUsed,
+      });
+
+      // Put leftover elements back on the front of the queue
+      if (leftoverElements.length > 0) {
+        contestSectionsLeftToLayOut = [
+          leftoverElements.map(({ contest }) => contest),
+          ...contestSectionsLeftToLayOut,
+        ];
+      }
+
+      // If there wasn't enough room left for any contests, go to the next page
+      if (height === 0) {
+        break;
+      }
+
+      const [sectionRectangle, sectionPositions] = ContestSection({
         election,
-        contests: contestsLeftToLayOut,
-        columnRowHeight,
-        gridRow: TIMING_MARKS_ROW_HEIGHT + headerAndInstructionsRowHeight,
+        contestColumns: columns.map((column) =>
+          column.map(({ contest }) => contest)
+        ),
+        height,
+        gridRow:
+          TIMING_MARKS_ROW_HEIGHT +
+          headerAndInstructionsRowHeight +
+          yToRow(heightUsed),
         gridColumn: 2,
         pageNumber,
       });
-    debug(
-      `Layed out ${contestsLeftToLayOut.length - restContests.length} contests`
-    );
-    contestsLeftToLayOut = restContests;
+
+      debug(
+        `Layed out ${contestSection.length - leftoverElements.length} contests`
+      );
+      contestObjects.push(sectionRectangle);
+      heightUsed += height;
+      gridPositions.push(...sectionPositions);
+    }
+
+    if (contestObjects.length === 0) {
+      contestObjects.push({
+        type: 'TextBox',
+        ...gridPosition({
+          row:
+            TIMING_MARKS_ROW_HEIGHT +
+            headerAndInstructionsRowHeight +
+            contestsRowHeight / 2,
+          column: GRID.columns / 2 - 7,
+        }),
+        width: gridWidth(15),
+        height: gridHeight(2),
+        textLines: ['This page intentionally left blank.'],
+        ...FontStyles.H2,
+      });
+    }
```

Comment:
This applies to the case where for whatever reason, we can't fit contests on a page but can fit it on another page. You scrolled over an example of this in show and tell.

Where in the code do we add "This page intentionally left blank." to the back page if it is empty? Does this case cover it? Don't understand.

Reply:
E.g. contests all covered by page 5, but there has to be a page 6 because sheet 3 will have a back.

Reply:
I don't think that case is covered currently. I think it's just a fully blank page. I think that's how our old HMPB worked so I just replicated that. But I could be remembering wrong.

Reply:
I thought our old HMPB ballots had footers and QR codes on back pages even if they were blank, but could be misremembering.

Reply:
I think you're right: https://github.com/votingworks/vxsuite/blob/274ad4a1c4b35fd25bab49ff4d0dd12512134b4a/apps/admin/frontend/src/components/hand_marked_paper_ballot.tsx#L211-L215

Reply:
Easy enough to fix! 640157520